dadguide/monster_index.py

```diff
@@ -1,34 +1,155 @@
-from token_mappings import *
+import re
 from collections import defaultdict
 
+from .token_mappings import *
+
 
 class MonsterIndex2:
     def __init__(self, monsters: 'List[DgMonster]', series: 'List[DgSeries]'):
         self.monsters = monsters
         self.series = series
-        self.index = defaultdict(list)
+        self.index = defaultdict(set)
         self._build_index()
 
+    def update(self, other: "Mapping[Any, set]"):
+        for k,v in other.items():
+            self.index[k].update(v)
+
     def _build_index(self):
-        self.index = defaultdict(list)
-        self.monster_name_index = self._build_monster_name_index(self.monsters)
-        self.series_name_index = self._build_series_name_index(self.series)
+        self.index = defaultdict(set)
+        self.update(self._build_monster_name_index(self.monsters))
+        self.update(self._build_color_index(self.monsters))
+        self.update(self._build_better_series_index(self.monsters))
+        self.update(self._build_evo_index(self.monsters))
+        self.update(self._build_etc_index(self.monsters))
+        #self.series_name_index = self._build_series_name_index(self.series)
 
     def _build_monster_name_index(self, monsters):
-        tokens = defaultdict(list)
+        tokens = defaultdict(set)
         for monster in monsters:
             for token in self._name_to_tokens(monster.name_en):
-                tokens[token.lower()].append(monster)
+                tokens[token.lower()].add(monster)
         return tokens
 
     def _build_series_name_index(self, series):
-        tokens = defaultdict(list)
+        tokens = defaultdict(set)
         for s in series:
             for token in self._name_to_tokens(s.name_en):
-                tokens[token.lower()].append(s)
+                tokens[token.lower()].add(s)
         return tokens
 
     def _name_to_tokens(self, name):
         name = re.sub(r'[\-+]', ' ', name.lower())
         name = re.sub(r'[^a-z0-9 ]', '', name)
         return [t.strip() for t in name.split() if t]
+
+    def _build_color_index(self, monsters):
+        tokens = defaultdict(set)
+
+        # Main Color
+        for c in COLOR_MAP:
```

Comment:
this outer loop seems unnecessary. cant you loop over monsters and look at its attr1.value and attr2.value?